### PR TITLE
Adding associated pod list screen for service details

### PR DESCRIPTION
**Pull request by rakol-ms**
Signed-off-by: Rajendra Kolli <rakol@microsoft.com>

svc_details

### CODE REVIEW

File: src/WebUI/Components/ServiceComponent.scss [hunk 1]
```diff
@@ -10,4 +10,24 @@
...
+.s-pod-list-content {
+    display: flex;
+    flex-direction: column;
+    justify-content: space-around;
+    margin-top: 10px; 
+}
+
+.s-pod-col-header {
+    color: $secondary-text;
+    font-size: $fontSizeM;
```
Comment (madhuv-msft) [inline src/WebUI/Components/ServiceComponent.scss]: do we need this
font-size: $fontSizeM;

File: src/WebUI/Components/ServiceComponent.scss [hunk 2]
```diff
@@ -10,4 +10,24 @@
 
 .s-de-heading {
     padding-left: 20px;
-}
\ No newline at end of file
+}
+
+.s-pod-list-content {
```
Comment (madhuv-msft) [inline src/WebUI/Components/ServiceComponent.scss]: .s-pod-list-content {
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    margin-top: 10px; 
}

.s-pod-col-header {
    color: $secondary-text;


if these are common settings we should set them in common place

File: src/WebUI/Components/ServiceComponent.tsx [hunk 3]
```diff
@@ -14,17 +14,44 @@ import * as Resources from "../Resources";
...
+            pods: []
+        };
+    }
+
     public render(): JSX.Element {
         return (
             <div className="service-main-content">
                 {this._getMainHeading()}
                 {this._getServiceDetails()}
+                {!!this.state.showAssociatedPods && this._getAssociatedPods()}
```
Comment (madhuv-msft) [inline src/WebUI/Components/ServiceComponent.tsx]: !!this.state.showAssociatedPods && --> we do not need this, as we should show column names when data is not available

File: src/WebUI/Components/ServiceComponent.tsx [hunk 4]
```diff
@@ -124,4 +151,117 @@ export class ServiceComponent extends BaseComponent<IServiceComponentProperties>
                 return renderSimpleCell(rowIndex, columnIndex, tableColumn, tableItem);
         }
     }
+
+    public componentDidMount(): void {
+        console.log("getting items");
+        this.props.podListingPromise.then(podList => {
+            podList.items &&
```
Comment (madhuv-msft) [inline src/WebUI/Components/ServiceComponent.tsx]: podList && podList.items &&

File: src/WebUI/Components/ServiceComponent.tsx [hunk 5]
```diff
@@ -124,4 +151,117 @@ export class ServiceComponent extends BaseComponent<IServiceComponentProperties>
...
+                    showAssociatedPods: true,
+                    pods: podList.items
+                });
+        }).catch(error => {
+            console.log(error);
+        });
+    }
+
+    private _getAssociatedPods(): JSX.Element | null {
+        if(this.state.showAssociatedPods) {
```
Comment (madhuv-msft) [inline src/WebUI/Components/ServiceComponent.tsx]: this check is not needed
if(this.state.showAssociatedPods) {

File: src/WebUI/Utils.ts [hunk 6]
```diff
@@ -25,4 +26,24 @@ export class Utils {
 
         return labelArray;
     }
+
+    public static generateEqualsConditionLabelSelector(labels: { [key: string]: string }): string {
+        console.log(labels);
+        let labelSelector: string = "";
+        if(labels) {
+            let keySet = Object.keys(labels);
```
Comment (madhuv-msft) [inline src/WebUI/Utils.ts]: const keySet

File: src/WebUI/Utils.ts [hunk 7]
```diff
@@ -25,4 +26,24 @@ export class Utils {
...
     }
+
+    public static generateEqualsConditionLabelSelector(labels: { [key: string]: string }): string {
+        console.log(labels);
+        let labelSelector: string = "";
+        if(labels) {
+            let keySet = Object.keys(labels);
+            keySet.forEach((key,index) => {
+                labelSelector = labelSelector.concat(format("{0}={1}", key, labels[key]))
+                if (index < keySet.length-1) labelSelector = labelSelector.concat(",");
```
Comment (madhuv-msft) [inline src/WebUI/Utils.ts]: if (index < keySet.length-1), not needed i guess

File: src/WebUI/Utils.ts [hunk 8]
```diff
@@ -25,4 +26,24 @@ export class Utils {
...
+            });
+        }
+        return labelSelector;
+    }
+
+    public static generatePodStatusProps(status:V1PodStatus): IStatusProps {
+        if(status.phase === "Running" || status.phase === "Scheduled"){
```
Comment (madhuv-msft) [inline src/WebUI/Utils.ts]: can we use contracts of the k8s contracts? instead of "Running" etc...

**Comment by rakol-ms** [comment]
@madhuv-msft Addressed review comments. PFA the screens for no pods and a svc with pods

empty_svc_details
svc_details

**Comment by rakol-ms** [comment]
orphanpod

**Comment by madhuv-msft** [comment]
On the latest UI:
we should remove "Orphan Pods" heading. in the same window, make "Pods" as "Pod"
In orphan replicaset window, remove heading of the replicaset and labels for the same. we can talk to PM to show labels as column. we can take that little later.
in this workloads page, "Pods" column should be renamed to "Pods running", this is for deploypment, replicaset, daemon, stateful set tables.

File: src/Contracts/Contracts.ts [hunk 9]
```diff
@@ -6,7 +6,7 @@
 import * as K8sTypes from "@kubernetes/client-node";
 
 export interface IKubeService {
-    getPods(): Promise<K8sTypes.V1PodList>;
+    getPods(labelSelector?:string): Promise<K8sTypes.V1PodList>;
```
Comment (madhuv-msft) [inline src/Contracts/Contracts.ts]: see if you can add a test for this?

File: src/WebUI/Components/KubeSummary.tsx [hunk 10]
```diff
@@ -175,6 +185,7 @@ export class KubeSummary extends BaseComponent<IKubeSummaryProps, IKubernetesCon
                     key={format("dc-{0}", this.state.namespace || "")}
                     onItemInvoked={this._onDeploymentItemInvoked}
                 />
+                {this.state.showOrphanPods && this.getOrphanPods()}
```
Comment (madhuv-msft) [inline src/WebUI/Components/KubeSummary.tsx]: should we move this deploymentscomponent, or rename deploymentscomponent to workloadscomponent?

File: src/WebUI/Utils.ts [hunk 11]
```diff
@@ -3,10 +3,22 @@
...
+import { V1ObjectMeta, V1PodStatus } from "@kubernetes/client-node";
 import { format } from "@uifabric/utilities/lib";
 import { ObservableArray } from "azure-devops-ui/Core/Observable";
 import { ILabelModel } from "azure-devops-ui/Label";
+import { Statuses, IStatusProps } from "azure-devops-ui/Status";
+
+/**
+ * https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
+ */
+export enum PodPhase {
```
Comment (madhuv-msft) [inline src/WebUI/Utils.ts]: if ant used any where we can make this as without export, if used, we can keep this in types.ts file

File: tests/Contracts/KubeServiceBase.test.ts [hunk 12]
```diff
@@ -41,4 +41,12 @@ describe("KubeServiceBase Tests", () => {
             expect(output).toBe(KubeResourceType.Services);
         });
     });
+
+    it("getPods calls with labelSelector as input",() => {
+        expect.assertions(1);
+        const labelSelector:string = "app=app";
+        return service.getPods(labelSelector).then(output =>{
```
Comment (madhuv-msft) [inline tests/Contracts/KubeServiceBase.test.ts]: we can add a case for labelSelector also by maksing changes to MockKubeService class

File: src/WebUI/Utils.ts [hunk 13]
```diff
@@ -25,4 +37,24 @@ export class Utils {
...
+        console.log(labels);
+        let labelSelector: string = "";
+        if(labels) {
+            const keySet = Object.keys(labels);
+            keySet.forEach((key,index) => {
+                labelSelector = labelSelector.concat(format("{0}={1}", key, labels[key]))
+                if (index < keySet.length-1) labelSelector = labelSelector.concat(",");
+            });
+        }
+        return labelSelector;
```
Comment (madhuv-msft) [inline src/WebUI/Utils.ts]: if we want a comma separated list, we can put the labels in to a string and finally can concat

**Pull request merged**